src/librbd/api/DiffIterate.cc

```diff
@@ -69,7 +69,7 @@ class C_DiffObject : public Context {
       m_image_ctx, io::IMAGE_DISPATCH_LAYER_INTERNAL_START,
       aio_comp, {{m_image_offset, m_image_length}},
       {m_diff_context.from_snap_id, m_diff_context.end_snap_id},
-      (m_diff_context.include_parent ?
+      ((m_diff_context.include_parent && m_diff_context.from_snap_id == 0) ?
         0 : io::LIST_SNAPS_FLAG_DISABLE_LIST_FROM_PARENT) |
```

Comment:
nit: may be it is time to introduce local `flags` variable, calculate its value and pass to `io::ImageDispatchSpec::create_list_snaps` to make the code more readable?

Reply:
Agreed.